### DISCUSSION TITLE
[Feature Request] Retrieve all badges / not awarded badges. And show progress towards obtaining ?

### DISCUSSION THREAD

**Discussion by vesper8** [maize-tech/laravel-badges#2] [Ideas]
First of all thanks for making this package! It's just what I was looking for.

In my project I want to have a page that displays all the possible achievements (badges) and they would all be greyed out or black and white, and the ones that have been achieved would be coloured. And you would also unlock rewards for the awarded badges (in terms of in-app currency).

Currently I don't see any way to retrieve a list of all badges, it seems I can only get the badges that a user has been awarded. So would it be possible to add a method/helper for retrieving all the possible badges?

Also.. for badges that require more than 1 of a thing.. such as needing 5 records to exist, would it be possible to return this.. so I could display that the current progress is 3 out of 5?

As for the ability to associate points to badges.. I get I can already manage this using the metadata but I wonder if it would make sense to have this kind of functionality baked in? Some food for thought.

Thanks for considering these requests! Cheers!

**Comment by riccardodallavia** [comment]
Hi @vesper8,

first of all, thank you for sharing your thoughts!

You can already retrieve a list of all available badges with a few lines of code:

`collect(\Maize\Badges\Support\Config::getBadgeClasses())->map(fn (string $badgeClass) => new $badgeClass);`

If working with RESTful paginated APIs you can also inject the given collection to a `LengthAwarePagination` with ease:

```
public function paginate(int $perPage): LengthAwarePaginator
{
    $page = Paginator::resolveCurrentPage();
    $items = $this>getItems();

    return new LengthAwarePaginator(
        items: $this->items->forPage($page, $perPage),
        total: $this->items->count(),
        perPage: $perPage,
        currentPage: $page,
        options: [
            'path' => Paginator::resolveCurrentPath(),
        ]
    );
}

protected function getItems(): \Illuminate\Support\Collection
{
    return collect(
        \Maize\Badges\Support\Config::getBadgeClasses()
    )->map(
        fn (string $badgeClass) => new $badgeClass
    );
}
```

The associated resource can also retrieve whether a badge has been achieved or not:

```
public function toArray(Request $request): array
{
    return [
        'name' => $this->getMetadata('name'),
        'description' => $this->getMetadata('description'),
        'is_progressable' => is_a($this->resource, ProgressableBadge::class),
        $this->mergeWhen(
            condition: is_a($this->resource, ProgressableBadge::class),
            value: fn () => [
                'progress' => min($this->getCurrent($request->user()), $this->getTotal()),
                'progress_total' => $this->getTotal(),
            ]
        ),
        'is_awarded' => $request->user()->badges->where('badge', $this->slug())->isNotEmpty(),
    ];
}
```

As per your suggestion:

> As for the ability to associate points to badges.. I get I can already manage this using the metadata but I wonder if it would make sense to have this kind of functionality baked in? Some food for thought.

I don't think we'll be adding a feature like this, because our goal is to keep the package simple and easy to extend using metadatas (where applicable) or by overriding the Badge class.
